### ISSUE TITLE
Sockeye transformer has different total number of trainable parameters from T2T Transformer

### ISSUE THREAD

**Issue by szhengac** [awslabs/sockeye#421]
I read your arxiv paper, and I found that the total number of trainable parameters of SOCKEYE transformer is 62,946,611 on EN→DE task while the number is 60,668,928 for T2T transformer. I wonder what contributes to this difference?

**Comment by fhieber** [comment]
Hi @szhengac, the parameter sizes of the transformer layers are the same. If I remember correctly the difference was due to the # of embedding parameters in Sockeye and T2T. Afaik, T2T splits the embedding matrix into multiple evenly sized shards. There might also be small differences in the vocabulary size (number of special symbols added in each toolkit etc.).

**Comment by szhengac** [comment]
@fhieber Thanks for your explanation. I would appreciate it if you could provide me more details (e.g., vocabulary size). Two million looks like a big difference simply due to different vocabulary sizes. For the embedding vector of size 512, it is equivalent to having 4,000 more tokens in the vocabulary of SOCKEYE  transformer.

**Comment by szhengac** [comment]
@fhieber Thanks for clearing it up. I tried sockeye_autopilot and I obtained a vocabulary of size 37135, which is different from 36659 in your log. Is this due to different frequency threshold? I used  preprocessed dataset downloaded from [https://github.com/tensorflow/nmt/blob/master/nmt/scripts/wmt16_en_de.sh]. This script produces a vocabulary of size 36548 without using frequency threshold.

**Comment by fhieber** [comment]
Note that our arxiv paper used WMT 17(!) data which is not the same as preprocessed dataset you are referring to.
You can find information on the experiments for the arxiv paper in the arxiv branch of this repository: https://github.com/awslabs/sockeye/tree/arxiv_1217/arxiv/code

**Comment by szhengac** [comment]
Oh, thanks, I just noted that.

**Comment by fhieber** [comment]
Closing this for now, but feel free to reopen if you have more questions.

**Comment by szhengac** [comment]
Hi @fhieber, I noted that the following preprocessing steps on train and test sets are performed in your arxiv paper:

`normalize-punctuation.perl `
`remove-non-printing-char.perl`

However, this is not done on the reference for computing BLEU score (in sacrebleu code). Would this cause some inconsistency? Thanks.

**Comment by mjpost** [comment]
Hi @szhengac—do you mean inconsistency in BLEU scores? We detokenized the outputs before scoring them with sacrebleu, so there should not be any problems across systems.

**Comment by szhengac** [comment]
Hi @mjpost, I didn't mean the inconsistency across systems. I see that the test set has been preprocessed by the steps I mentioned above before being fed into the model, but the reference in sacreble is not preprocessed by these steps. So even the ground truth translation cannot have a 100 bleu score, because the detokenizer only reverses the tokenier but not including normalize-punctuation.perl and remove-non-printing-char.perl

**Comment by mjpost** [comment]
Yes, this is true in general that tokenization is not perfectly invertable. However, sacrebleu's approach of "hiding" the reference and applying a consistent tokenization is preferable to the situation where everyone applies their own tokenization to the reference, because in the latter situation, the scores cannot be compared across papers.

**Comment by szhengac** [comment]
Yes, I understand that. However, then what is the purpose to use that preprocessing pipeline if it is not applied on the reference?  In my testing, this can cause a difference of about 1.9 BLEU score.

**Comment by mjpost** [comment]
Preprocessing is applied at training time to regularize the training data (e.g., you want "food" and "food," not to become separate tokens). Preprocessing *is* applied to the reference.

**Comment by szhengac** [comment]
Could you give me the link to the code for "Preprocessing is applied to the reference." I know that the preprocessing is applied to the test set in [prepare_devtest.sh](https://github.com/awslabs/sockeye/blob/arxiv_1217/arxiv/code/prepare_devtest.sh). However, it seems that the preprocessing is not applied to the reference for the BLEU evaluation. For example, from what I understand, newstest2014.bpe.en (including preprocessing on en-de.en) is used to produce translation, which is used to compute the BLEU against unpreprocessed en-de.de.

**Comment by mjpost** [comment]
The tokenization is applied inside [sacrebleu](https://github.com/mjpost/sacreBLEU/blob/3ac82093f48d067157446afc81a7834294d148b4/sacrebleu.py#L495). The default tokenization (which can be changed via the `-tok TYPE` flag) is "v13a", which is chosen to be compatible with the [WMT scoring script](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/mteval-v13a.pl) used to compute scores on [the matrix](http://matrix.statmt.org).

**Comment by szhengac** [comment]
I may not make my question very clear. For preprocessing, I only refer to `normalize-punctuation.perl` and
`remove-non-printing-char.perl`. As the `detokenizer.perl` only reverses the `tokenizer.perl` but not including `normalize-punctuation.perl` and `remove-non-printing-char.perl`, I think it is more consistent to perform tokenization (e.g., v13a) on following tweaked reference:

`cat en-de.de | normalize-punctuation.perl -l de | remove-non-printing-char.perl  > en-de.de.tweaked`

**Comment by mjpost** [comment]
This is probably true that this would be more consistent. But 

- while this would increase BLEU scores, it is not a very meaningful increase 
- you would not be able to compare BLEU scores against other groups who have not applied these additional preprocessing to the reference

**Comment by szhengac** [comment]
Yes, that is right. So my concern is that these additional preprocessing may make the BLEU worse compared to the case where the training and decoding are performed on the datasets without these additional preprocessing, because the latter one is more consistent with the evaluation.

**Comment by mjpost** [comment]
Yes, that's possible. People submitting to evaluations like WMT have dealt with this for a long time. One recommendation WMT makes is [the following](http://statmt.org/wmt18/translation-task.html) (see section "TEST SET SUBMISSION"). 

> Punctuation in the official test sets will be encoded with ASCII characters (not complex Unicode characters) as much as possible. You may want to normalize your system's output before submission.

I have seen this make a +3 point increase in BLEU score for a particular system that used smart quotes instead of the normalized quotes found in the test set reference.

**Comment by szhengac** [comment]
This is a big difference. Have you guys tried reproducing the results on Attention Is All You Need paper? As far as I know, their results were evaluated on tweaked reference.

**Comment by mjpost** [comment]
Yes. In fact, the difficulties we had repeating their results (due largely to discrepancies stemming from ther tweaked reference, as you put it) were part of the motivation behind sacrebleu and the [sacrebleu paper](http://arxiv.org/pdf/1804.08771v1).
